nerfstudio/models/gaussian_splatting.py

```diff
@@ -0,0 +1,653 @@
+# Copyright 2022 the Regents of the University of California, Nerfstudio Team and contributors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+NeRF implementation that combines many recent advancements.
+"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Dict, List, Optional, Tuple, Type, Union
+import torch
+from nerfstudio.data.scene_box import OrientedBox
+from copy import deepcopy
+from nerfstudio.cameras.rays import RayBundle
+from torch.nn import Parameter
+from torchmetrics.image import PeakSignalNoiseRatio
+from torchmetrics.image.lpip import LearnedPerceptualImagePatchSimilarity
+from nerfstudio.cameras.cameras import Cameras
+from gsplat._torch_impl import quat_to_rotmat
+from nerfstudio.engine.callbacks import TrainingCallback, TrainingCallbackAttributes, TrainingCallbackLocation
+from nerfstudio.engine.optimizers import Optimizers
+from nerfstudio.models.base_model import Model, ModelConfig
+import math
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+import viser.transforms as vtf
+
+
+from gsplat.rasterize import RasterizeGaussians
+from gsplat.project_gaussians import ProjectGaussians
+from gsplat.sh import SphericalHarmonics, num_sh_bases
+
+
+def random_quat_tensor(N, **kwargs):
+    u = torch.rand(N, **kwargs)
+    v = torch.rand(N, **kwargs)
+    w = torch.rand(N, **kwargs)
+    return torch.stack(
+        [
+            torch.sqrt(1 - u) * torch.sin(2 * math.pi * v),
+            torch.sqrt(1 - u) * torch.cos(2 * math.pi * v),
+            torch.sqrt(u) * torch.sin(2 * math.pi * w),
+            torch.sqrt(u) * torch.sin(2 * math.pi * w),
+        ],
+        dim=-1,
+    )
+
+
+def RGB2SH(rgb):
+    C0 = 0.28209479177387814
+    return (rgb - 0.5) / C0
+
+
+def projection_matrix(znear, zfar, fovx, fovy, **kwargs):
+    t = znear * math.tan(0.5 * fovy)
+    b = -t
+    r = znear * math.tan(0.5 * fovx)
+    l = -r
+    n = znear
+    f = zfar
+    return torch.tensor(
+        [
+            [2 * n / (r - l), 0.0, (r + l) / (r - l), 0.0],
+            [0.0, 2 * n / (t - b), (t + b) / (t - b), 0.0],
+            [0.0, 0.0, (f + n) / (f - n), -1.0 * f * n / (f - n)],
```

Comment:
Is there a mistake in the projmat[2,2]? 
refer to reference code: https://github.com/graphdeco-inria/gaussian-splatting/blob/414b553ef1d3032ea634b7b7a1d121173f36592c/utils/graphics_utils.py#L69

Reply:
I guess there is some difference to [OpenGL](http://www.songho.ca/opengl/gl_projectionmatrix.html) convention too. Wonder which is correct now.

Reply:
I think the difference in (f+n) and just (f) is negligible probably. We can leave it as is.

Reply:
https://github.com/nerfstudio-project/nerfstudio/pull/2521/files/40c05be7a60bde25adc28cd281bb9d742c2a77e9#diff-fa59b1578ecffb1d76c0a43bec7a2a12217696ce7cb3632aba6af5731a881459R505   well the znear and zfar right now are hard-coded

( which is a bit confusing because the model inherits collider params ... https://github.com/nerfstudio-project/nerfstudio/blob/8a3e24738ba7b31139418e109896915a69bfb6a4/nerfstudio/models/base_model.py#L48   )

maybe the overall question is worth returning to later to see if the viser webgl viewer https://github.com/nerfstudio-project/viser/pull/110/files  ends up consistent with images rendered thru nerfstudio

seems at the least znear and zfar should be user configurable instead of hardcoded...  those are effectively hyperparameters, if not also the whole P matrix used ... neither gsplat back-ends use OpenGL render-to-texture tho at least.